bindings/src/test/java/io/pravega/storage/extendeds3/S3Mock.java

```diff
@@ -0,0 +1,315 @@
+            deleteObject(request.getBucketName(), obj.getKey());
+        }
+        return new DeleteObjectsResult();
+    }
+
+    public ListObjectsResult listObjects(String bucketName, String prefix) {
+        ListObjectsResult result = new ListObjectsResult();
+        List<S3Object> list;
+        String objectPrefix = getObjectName(bucketName, prefix);
+        String bucketPrefix = getObjectName(bucketName, "");
+        synchronized (this.objects) {
+            list = this.objects.entrySet().stream()
+                    .filter(e -> e.getKey().startsWith(objectPrefix))
+                    .map(e -> {
+                        S3Object object = new S3Object();
+                        object.setKey(e.getKey().substring(bucketPrefix.length()));
+                        object.setSize((long) e.getValue().content.getLength());
+                        object.setLastModified(new Date(System.currentTimeMillis()));
+                        return object;
+                    })
+                    .collect(Collectors.toList());
+        }
+        result.setObjects(list);
+        result.setMaxKeys(list.size());
+        result.setBucketName(bucketName);
+        result.setPrefix(prefix);
+        return result;
+    }
+
+    public S3ObjectMetadata getObjectMetadata(String bucketName, String key) {
+        String objectName = getObjectName(bucketName, key);
+        S3ObjectMetadata metadata = new S3ObjectMetadata();
+        synchronized (this.objects) {
+            ObjectData od = this.objects.get(objectName);
+            if (od == null) {
+                throw new S3Exception("NoSuchKey", 404, "NoSuchKey", "");
+            }
+            metadata.setContentLength((long) od.content.getLength());
+            metadata.setLastModified(new Date(System.currentTimeMillis()));
+        }
+        return metadata;
+    }
+
+    public InputStream readObjectStream(String bucketName, String key, Range range) {
+        String objectName = getObjectName(bucketName, key);
+        synchronized (this.objects) {
+            ObjectData od = this.objects.get(objectName);
+            if (od == null) {
+                throw new S3Exception("NoSuchKey", 404, "NoSuchKey", "");
+            }
+
+            final int offset = (int) (long) range.getFirst();
+            final int length = (int) (range.getLast() - range.getFirst() + 1);
+            try {
+                return od.content.slice(offset, length).getReader();
+            } catch (Exception ex) {
+                throw new S3Exception("InvalidRange", org.apache.http.HttpStatus.SC_REQUESTED_RANGE_NOT_SATISFIABLE, "InvalidRange", "");
+            }
+        }
+    }
+
+    public String initiateMultipartUpload(String bucketName, String key) {
+        String objectName = getObjectName(bucketName, key);
+        synchronized (this.objects) {
+            this.multipartUploads.put(objectName, new HashMap<>());
+            return Integer.toString(this.multipartUploads.size());
```

Comment:
where is this string used?

I believe you can have multiple MPU requests against the same object.

Reply:
From what I could tell, it returns some identifier that is then used to reference this MPU.

This implementation was taken from the defunct `S3ProxyFileSystemImpl.java`.

Reply:
I changed to a monotonically increasing AtomicLong so we don't ever have collisions.